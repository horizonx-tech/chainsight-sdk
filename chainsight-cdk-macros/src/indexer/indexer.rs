use proc_macro::TokenStream;
use quote::quote;
use syn::{
    parse::{Parse, ParseStream},
    punctuated::Punctuated,
    Ident, Result, Token,
};

pub struct Args {
    contract_name: String,
    event_name: String,
    args: Punctuated<NamedField, Token![,]>,
    address: String,
}
struct NamedField {
    name: Ident,
    _colon_token: Token![:],
    ty: syn::Type,
}
impl Parse for NamedField {
    fn parse(input: ParseStream) -> Result<Self> {
        Ok(NamedField {
            name: input.parse()?,
            _colon_token: input.parse()?,
            ty: input.parse()?,
        })
    }
}
impl Parse for Args {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let contract_name: syn::LitStr = input.parse()?;
        input.parse::<syn::Token![,]>()?;
        let event_name: syn::LitStr = input.parse()?;
        input.parse::<syn::Token![,]>()?;
        let args: Punctuated<NamedField, Token![,]> = Punctuated::parse_terminated(input)?;
        input.parse::<syn::Token![,]>()?;
        let address: syn::LitStr = input.parse()?;
        Ok(Args {
            contract_name: contract_name.value(),
            event_name: event_name.value(),
            args,
            address: address.value(),
        })
    }
}

pub fn gen(input: TokenStream) -> TokenStream {
    let Args {
        contract_name,
        event_name,
        args,
        address,
    } = syn::parse_macro_input!(input as Args);
    let struct_name = Ident::new(&format!("{}Event", event_name), event_name.span());

    quote! {}.into()
}
