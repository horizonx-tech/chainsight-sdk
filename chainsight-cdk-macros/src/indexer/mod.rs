pub mod indexer;
pub use indexer::*;
