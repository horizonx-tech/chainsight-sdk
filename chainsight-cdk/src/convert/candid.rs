use std::{cmp::max, fs, ops::Deref, path::Path};

use anyhow::Ok;
use candid::{
    bindings::rust::{compile, Target},
    check_prog,
    types::{Type, TypeInner},
    IDLProg, TypeEnv,
};
use regex::Regex;

#[derive(Debug, Clone)]
pub struct CanisterMethodIdentifier {
    pub identifier: String,
    type_env: TypeEnv,
}
impl CanisterMethodIdentifier {
    pub const REQUEST_ARGS_TYPE_NAME: &'static str = "RequestArgsType";
    pub const RESPONSE_TYPE_NAME: &'static str = "ResponseType";
    pub const SUFFIX_IN_DID: &'static str = "InDid";

    pub fn new(s: &str) -> anyhow::Result<Self> {
        Self::new_internal(s, None)
    }

    pub fn new_with_did(s: &str, dependended_did: String) -> anyhow::Result<Self> {
        // NOTE: If the reserved Type specified in the .did to import is used, a Compile error will occur due to duplication.
        let (s, dependended_did) = avoid_using_reserved_types(s, &dependended_did);
        Self::new_internal(&s, Some(dependended_did))
    }

    fn new_internal(s: &str, dependended_did: Option<String>) -> anyhow::Result<Self> {
        let (identifier, args_ty, response_ty) = extract_elements(s)?;
        let did: String = Self::generate_did(&args_ty, &response_ty);

        let ast: IDLProg = if let Some(base_did) = dependended_did {
            format!("{}\n\n{}", base_did, did)
        } else {
            did.to_string()
        }
        .parse()?;
        let mut type_env = TypeEnv::new();
        let _ = check_prog(&mut type_env, &ast);

        Ok(Self {
            identifier,
            type_env,
        })
    }

    pub fn compile(&self) -> anyhow::Result<String> {
        anyhow::ensure!(self.compilable(), "Not compilable IDLProg");

        let mut config = candid::bindings::rust::Config::new();
        // // Update the structure derive to chainsight's own settings
        config.type_attributes = "#[derive(Clone, Debug, candid :: CandidType, candid :: Deserialize, serde :: Serialize, chainsight_cdk_macros :: StableMemoryStorable)]".to_string();
        config.target = Target::CanisterStub;
        let contents = compile(&config, &self.type_env, &None);

        let mut lines = contents
            .lines()
            // Delete initial 'use' declarations
            .skip(4)
            // Delete comment lines and blank lines
            .filter(|line| !(line.starts_with("//") || line.is_empty()))
            .map(|line| {
                // convert icp's own Nat and Int types to rust native type
                // NOTE: num-traits can be used, but is not used to reduce dependencies
                //  https://forum.dfinity.org/t/candid-nat-to-u128/16016
                //  https://discord.com/channels/748416164832608337/872791506853978142/1162494173933481984
                // NOTE: when ready to convert to u128/i128, consider with EthAbiEncoder's Encoder trait
                if line.contains("candid::Nat") {
                    return line.replace("candid::Nat", "u128");
                }
                if line.contains("candid::Int") {
                    return line.replace("candid::Int", "i128");
                }
                line.to_string()
            })
            .collect::<Vec<_>>();
        lines.insert(0, "#![allow(dead_code, unused_imports)]".to_string());
        lines.insert(
            1,
            "use candid::{self, CandidType, Deserialize, Principal, Encode, Decode};".to_string(),
        );
        Ok(lines.join("\n"))
    }

    pub fn get_types(&self) -> (Option<&Type>, Option<&Type>) {
        (
            self.find_type(Self::REQUEST_ARGS_TYPE_NAME),
            self.find_type(Self::RESPONSE_TYPE_NAME),
        )
    }

    pub fn get_type(&self, key: &str) -> Option<&Type> {
        self.find_type(key)
    }

    fn compilable(&self) -> bool {
        let (args_ty, response_ty) = self.get_types();
        let not_compilable_type = &TypeInner::Unknown;

        if args_ty.is_some_and(|ty| ty.deref().eq(not_compilable_type)) {
            return false;
        }
        if response_ty.is_some_and(|ty| ty.deref().eq(not_compilable_type)) {
            return false;
        }

        true
    }

    fn generate_did(args_ty: &str, response_ty: &str) -> String {
        let args_ty_did = if args_ty.is_empty() {
            "".to_string()
        } else {
            generate_did_type(Self::REQUEST_ARGS_TYPE_NAME, args_ty)
        };
        let response_ty_did = if response_ty.is_empty() {
            "".to_string()
        } else {
            generate_did_type(Self::RESPONSE_TYPE_NAME, response_ty)
        };
        format!("{}\n{}", args_ty_did, response_ty_did)
    }

    fn find_type(&self, key: &str) -> Option<&Type> {
        let ty = self.type_env.find_type(key);
        ty.ok()
    }
}

fn generate_did_type(key: &str, value: &str) -> String {
    format!("type {} = {};", key, value)
}

fn extract_elements(s: &str) -> anyhow::Result<(String, String, String)> {
    let (identifier, remains) = s
        .split_once(':')
        .expect("Invalid canister method identifier");
    let (args_ty, response_ty) = remains
        .split_once("->")
        .expect("Invalid canister method identifier");

    let trim_type_str = |s: &str| {
        let trimed = s.trim();
        let removed_brackets = trimed.trim_matches(|c| c == '(' || c == ')');
        removed_brackets.trim().to_string()
    };

    Ok((
        identifier.trim().to_string(),
        trim_type_str(args_ty),
        trim_type_str(response_ty),
    ))
}

fn avoid_using_reserved_types(s: &str, did: &str) -> (String, String) {
    let req_ty = CanisterMethodIdentifier::REQUEST_ARGS_TYPE_NAME;
    let res_ty = CanisterMethodIdentifier::RESPONSE_TYPE_NAME;
    let base_suffix = CanisterMethodIdentifier::SUFFIX_IN_DID;

    // Check maximum value of suffix for reserved type
    let pattern = format!(
        r"(?P<name>({}|{}))_{}_(?P<suffix_num>\d+)",
        req_ty, res_ty, base_suffix
    );
    let re = Regex::new(&pattern).expect("Invalid regex pattern");
    let mut max_number = 0;
    for line in did.lines() {
        for cap in re.captures_iter(line) {
            let suffix_num = cap.name("suffix_num").unwrap().as_str();
            if let core::result::Result::Ok(number) = suffix_num.parse::<u32>() {
                max_number = max(max_number, number);
            }
        }
    }
    max_number += 1;

    // Replace reserved type (did)
    let pattern = format!(r" (?P<name>({}|{}))(?P<last_char>(;|,| ))", req_ty, res_ty);
    let re = Regex::new(&pattern).expect("Invalid regex pattern");
    let replace_reserved_type = |s: &str| {
        let replaced_s = re.replace_all(s, |caps: &regex::Captures| {
            let name = caps.name("name").unwrap().as_str();
            let last_char = caps.name("last_char").unwrap().as_str();
            format!(" {}_{}_{}{}", name, base_suffix, max_number, last_char)
        });
        replaced_s.to_string()
    };
    let replaced_did_lines = did.lines().map(replace_reserved_type).collect::<Vec<_>>();

    // Replace reserved type (identifier)
    let pattern = format!(
        r"(?P<first_char>(\(| ))(?P<name>({}|{}))(?P<last_char>(\)|;|,| ))",
        req_ty, res_ty
    );
    let re = Regex::new(&pattern).expect("Invalid regex pattern");
    let replaced_s = re.replace_all(s, |caps: &regex::Captures| {
        let first_char = caps.name("first_char").unwrap().as_str();
        let name = caps.name("name").unwrap().as_str();
        let last_char = caps.name("last_char").unwrap().as_str();
        format!(
            "{}{}_{}_{}{}",
            first_char, name, base_suffix, max_number, last_char
        )
    });

    (replaced_s.to_string(), replaced_did_lines.join("\n"))
}

// Read .did and remove 'service' section
pub fn read_did_to_string_without_service<P: AsRef<Path>>(path: P) -> anyhow::Result<String> {
    let mut data = fs::read_to_string(path)?;
    Ok(exclude_service_from_did_string(&mut data))
}
fn exclude_service_from_did_string(data: &mut String) -> String {
    // remove 'service' section
    if let Some(service_start_idx) = data.find("service : {") {
        data.truncate(service_start_idx);
    }

    data.trim().to_owned()
}

#[cfg(test)]
mod tests {
    use super::*;

    use insta::assert_snapshot;

    const TEST_IDENTS: &'static [(&'static str, &'static str); 6] = &[
        (&"normal", &"update_value : (nat64) -> (text)"),
        (&"normal_nat", &"use_nat : (nat32) -> (nat)"),
        (&"normal_int", &"use_nat : (int32) -> (int)"),
        (&"no_args", &"get_value : () -> (text)"),
        (
            &"record",
            &"get_snapshot : (nat64) -> (record { value : text; timestamp : nat64 })",
        ),
        (
            &"nested_record",
            &"get_last_snapshot : () -> (record { dai : record { usd : float64 } })",
        ),
    ];

    const TEST_IDENTS_WITH_DID: &'static [(&'static str, &'static str, &'static str); 5] = &[
        (
            &"single",
            &"get_snapshot : (nat64) -> (Snapshot)",
            &"type Snapshot = record { value : text; timestamp : nat64 };",
        ),
        (
            &"multiple",
            &"get_snapshot : (nat64) -> (Snapshot)",
            &r#"type CurrencyValue = record { usd : float64 };
type Snapshot = record { value : SnapshotValue; timestamp : nat64 };
type SnapshotValue = record { dai : CurrencyValue };
"#,
        ),
        (
            &"with_variant",
            &"init_in : (Env) -> (Result)",
            &r#"type Env = variant { Production; Test; LocalDevelopment };
type InitError = variant {
  InvalidDestination : text;
  InvalidPrincipal : principal;
  InvalidContent : text;
  InvalidRequest : text;
};
type Result = variant { Ok; Err : InitError };
"#,
        ),
        (
            &"with_vec",
            &"get_sources : () -> (vec Sources)",
            &r#"type HttpsSnapshotIndexerSourceAttrs = record {
  queries : vec record { text; text };
};
type SourceType = variant { evm; https; chainsight };
type Sources = record {
  source : text;
  interval_sec : opt nat32;
  attributes : HttpsSnapshotIndexerSourceAttrs;
  source_type : SourceType;
};"#,
        ),
        (
            &"with_reserved_type",
            &"get_snapshot : (RequestArgsType) -> (ResponseType)",
            &r#"type RequestArgsType = nat64;
type ResponseType = text;"#,
        ),
    ];

    #[test]
    fn test_compile() {
        for (label, s) in TEST_IDENTS {
            let ident = CanisterMethodIdentifier::new(s).expect("Failed to parse");
            let compiled = ident.compile().unwrap();
            assert_snapshot!(format!("compile__{}", label.to_string()), compiled);
        }
    }

    #[test]
    fn test_compile_with_depended_did() {
        for (label, s, did) in TEST_IDENTS_WITH_DID {
            let ident = CanisterMethodIdentifier::new_with_did(s, did.to_string())
                .expect("Failed to parse");
            let compiled = ident.compile().unwrap();
            assert_snapshot!(
                format!("compile_with_depended_did__{}", label.to_string()),
                compiled
            );
        }
    }

    #[test]
    fn test_compile_when_not_compilable() {
        let s = "get_snapshot : (Input) -> (Output)";
        let ident = CanisterMethodIdentifier::new(s).expect("Failed to parse");
        assert_eq!(
            ident.compile().err().unwrap().to_string(),
            "Not compilable IDLProg"
        );
    }

    #[test]
    fn test_compilable() {
        // compilable
        for (_, s) in TEST_IDENTS {
            let ident = CanisterMethodIdentifier::new(s).expect("Failed to parse");
            assert!(ident.compilable());
        }
        for (_, s, did) in TEST_IDENTS_WITH_DID {
            let ident = CanisterMethodIdentifier::new_with_did(s, did.to_string())
                .expect("Failed to parse");
            assert!(ident.compilable());
        }

        // not compilable
        let not_compilable = vec![
            "get_snapshot : (Input) -> (Output)",
            "get_snapshot : (nat64) -> (Snapshot)",
            "get_snapshot : (Snapshot) -> (nat64)",
        ];
        for s in not_compilable {
            let ident = CanisterMethodIdentifier::new(s).expect("Failed to parse");
            assert!(!ident.compilable());
        }

        let not_compilable_with_did = vec![
            (
                "get_snapshot : (Snapshot) -> (Snapshot)",
                "type Snapshot_1 = record { value : text; timestamp : nat64 };",
            ),
            (
                "get_snapshot : (nat64) -> (Snapshot)",
                "type Snapshot_1 = record { value : text; timestamp : nat64 };",
            ),
            (
                "get_snapshot : (Snapshot) -> (nat64)",
                "type Snapshot_1 = record { value : text; timestamp : nat64 };",
            ),
        ];
        for (s, did) in not_compilable_with_did {
            let ident = CanisterMethodIdentifier::new_with_did(s, did.to_string())
                .expect("Failed to parse");
            assert!(!ident.compilable());
        }
    }

    #[test]
    fn test_generate_did() {
        let key = "Account";
        let value = "record { owner : principal; subaccount : opt blob }";
        let did = generate_did_type(key, value);
        assert_snapshot!(did);
    }

    #[test]
    fn test_extract_elements() {
        for (label, s) in TEST_IDENTS {
            let result = extract_elements(s).expect("Failed to parse");
            assert_snapshot!(
                format!("extract_elements__{}", label.to_string()),
                format!("{:#?}", result)
            );
        }
    }

    #[test]
    fn test_extract_elements_multi_nested_record() {
        let s =
            "icrc2_allowance : (record { account : record { owner : principal; subaccount : opt blob }; spender : record { owner : principal; subaccount : opt blob } }) -> (record { allowance : nat; expires_at : opt nat64 })";
        let (identifier, args_ty, response_ty) = extract_elements(s).expect("Failed to parse");
        assert_eq!(identifier, "icrc2_allowance");
        assert_eq!(args_ty, "record { account : record { owner : principal; subaccount : opt blob }; spender : record { owner : principal; subaccount : opt blob } }");
        assert_eq!(
            response_ty,
            "record { allowance : nat; expires_at : opt nat64 }"
        );
    }

    #[test]
    fn test_exclude_service_from_did_string() {
        let expected = r#"type List = opt record { head: int; tail: List };
type byte = nat8;".to_string()"#;
        let mut actual = format!("{}\n", expected)
            + r#"service : {
  f : (byte, int, nat, int8) -> (List);
  g : (List) -> (int) query;
}"#;

        assert_eq!(exclude_service_from_did_string(&mut actual), expected);
    }
}
