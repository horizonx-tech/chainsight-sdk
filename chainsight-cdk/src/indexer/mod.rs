mod indexer;
pub use indexer::*;
