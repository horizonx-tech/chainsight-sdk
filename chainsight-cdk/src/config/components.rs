use std::collections::HashMap;

use anyhow::bail;
use lazy_static::lazy_static;
use regex::Regex;

use crate::indexer::IndexingConfig;

lazy_static! {
    static ref MAPPING_CANDID_TY: HashMap<&'static str, &'static str> = [
        ("text", "String"),
        // ("blob", "&[u8]"),
        ("nat", "u128"),
        ("int", "i128"),
        ("nat8", "u8"),
        ("nat16", "u16"),
        ("nat32", "u32"),
        ("nat64", "u64"),
        ("int8", "i8"),
        ("int16", "i16"),
        ("int32", "i32"),
        ("int64", "i64"),
        ("float32", "f32"),
        ("float64", "f64"),
        ("bool", "bool"),
        // ("null", "()"),
    ].iter().cloned().collect();

    static ref REGEX_CANDID_FUNC: Regex = Regex::new(r"(?P<identifier>\w+)\s*:\s*\((?P<params>.*?)\)\s*(->\s*\((?P<return>.*?)\))?").unwrap();
    static ref REGEX_VECTOR: Regex = Regex::new(r"vec\s(?P<type>\w+)").unwrap();
    static ref REGEX_TUPLE: Regex = Regex::new(r"record\s\{\s(?P<items>(\w+(;\s|))+)\s\}").unwrap();
    static ref REGEX_STRUCT: Regex = Regex::new(r"(?P<field>\w+)\s*:\s*(?P<type>\w+)").unwrap();

    static ref REGEX_MULTIPLE_RECORD_TYPE: Regex = Regex::new(r"record\s*\{").unwrap();
}

#[derive(serde::Deserialize, serde::Serialize)]
pub struct CommonConfig {
    pub monitor_duration: u32,
    pub canister_name: String,
}

#[derive(serde::Deserialize, serde::Serialize)]
pub struct AlgorithmIndexerInput {
    pub method_name: String,
    pub response_type: String,
    pub source_type: AlgorithmInputType,
}

#[derive(Clone, Debug, PartialEq, serde::Deserialize, serde::Serialize)]
pub enum AlgorithmInputType {
    #[serde(rename = "event_indexer")]
    EventIndexer,
    #[serde(rename = "key_value")]
    KeyValue,
    #[serde(rename = "key_values")]
    KeyValues,
}

impl Default for AlgorithmIndexerInput {
    fn default() -> Self {
        Self {
            method_name: "get_list".to_string(),
            response_type: "String".to_string(),
            source_type: AlgorithmInputType::EventIndexer,
        }
    }
}

impl Default for CommonConfig {
    fn default() -> Self {
        Self {
            monitor_duration: 3600,
            canister_name: "example_canister".to_string(),
        }
    }
}

#[derive(Default, serde::Serialize, serde::Deserialize)]
pub struct AlgorithmIndexerConfig {
    pub common: CommonConfig,
    pub indexing: IndexingConfig,
    pub input: AlgorithmIndexerInput,
}

#[derive(Default, serde::Serialize, serde::Deserialize)]
pub struct AlgorithmLensConfig {
    pub common: CommonConfig,
    pub target_count: usize,
}

#[derive(Default, serde::Serialize, serde::Deserialize)]
pub struct EventIndexerConfig {
    pub common: CommonConfig,
    pub def: EventIndexerEventDefinition,
}
#[derive(Default, Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct EventIndexerEventDefinition {
    pub identifier: String,
    pub abi_file_path: String,
}

#[derive(serde::Serialize, serde::Deserialize)]
pub struct RelayerConfig {
    pub common: CommonConfig,
    pub destination: String,
    pub oracle_type: String,
    pub method_name: String,
    pub canister_method_value_type: CanisterMethodValueType,
    pub abi_file_path: String,
    pub lens_targets: Option<LensTargets>,
}
#[derive(Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct LensTargets {
    pub identifiers: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
pub enum CanisterMethodValueType {
    Scalar(String, bool),                // struct name, is_scalar
    Tuple(Vec<(String, bool)>),          // struct_name, is_scalar
    Struct(Vec<(String, String, bool)>), // temp: Only non-nested `record` are supported.
    Vector(String, bool),                // struct_name, is_scalar
}
#[derive(Debug, Clone, PartialEq)]
pub struct CanisterMethodIdentifier {
    pub identifier: String,
    pub params: Vec<String>,
    pub return_value: CanisterMethodValueType,
}

impl CanisterMethodIdentifier {
    pub fn parse_from_str(s: &str) -> anyhow::Result<Self> {
        let captures = REGEX_CANDID_FUNC.captures(s).ok_or(anyhow::anyhow!(
            "method.identifier does not satisfy the supported expression: {}, supplied={}",
            REGEX_CANDID_FUNC.to_string(),
            s
        ))?;

        let identifier = captures.name("identifier").unwrap().as_str().to_string();

        let params_str = captures.name("params").unwrap().as_str();
        let params_result: Vec<String> = if params_str.is_empty() {
            vec![]
        } else {
            params_str
                .split(',')
                .map(|s| convert_type_from_candid_type(s.trim()).0)
                .collect()
        };
        let params = params_result;

        let return_value_str = captures.name("return").unwrap().as_str();
        let return_value = Self::parse_return_value(return_value_str)?;

        Ok(CanisterMethodIdentifier {
            identifier,
            params,
            return_value,
        })
    }

    fn parse_return_value(s: &str) -> anyhow::Result<CanisterMethodValueType> {
        let record_type_count = REGEX_MULTIPLE_RECORD_TYPE.find_iter(s).count();
        if record_type_count >= 2 {
            bail!(
                "Sorry, Currently nested `record` types are not supported. This will be supported in the future.\nTarget literal = {}",
                s
            ); // TODO: Support nested `record` types.
        }
        // vector
        if s.starts_with("vec") {
            let captures = REGEX_VECTOR.captures(s);
            if let Some(captures_value) = captures {
                let ty = captures_value.name("type").unwrap().as_str();
                let val = convert_type_from_candid_type(ty);
                return Ok(CanisterMethodValueType::Vector(val.0, val.1));
            }
            bail!("Invalid candid's result types:{}", s);
        }

        // Scalar
        if !s.starts_with("record") {
            let val = convert_type_from_candid_type(s);
            return Ok(CanisterMethodValueType::Scalar(val.0, val.1));
        }

        // Tuple
        let captures = REGEX_TUPLE.captures(s);
        if let Some(captures_value) = captures {
            let items = captures_value.name("items").unwrap().as_str();
            let tuple_result: Vec<(String, bool)> = items
                .split(';')
                .map(|s| convert_type_from_candid_type(s.trim()))
                .collect();
            let tuple = tuple_result;
            return Ok(CanisterMethodValueType::Tuple(tuple));
        }

        // Struct
        let items = REGEX_STRUCT.captures_iter(s);
        let mut struct_items = vec![];
        for cap in items {
            let field = cap.name("field").unwrap().as_str().to_string();
            let ty = convert_type_from_candid_type(cap.name("type").unwrap().as_str());
            struct_items.push((field, ty.0, ty.1));
        }
        if struct_items.is_empty() {
            bail!("Invalid candid's result types: {}", s);
        }
        Ok(CanisterMethodValueType::Struct(struct_items))
    }
}

fn convert_type_from_candid_type(s: &str) -> (String, bool) {
    // ref: https://internetcomputer.org/docs/current/references/candid-ref
    let ty_str = MAPPING_CANDID_TY.get(&s);
    if let Some(ty_str) = ty_str {
        return (ty_str.to_string(), true);
    }
    (s.to_string(), false)
}
