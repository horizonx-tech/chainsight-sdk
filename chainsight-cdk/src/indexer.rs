use std::collections::HashMap;

use async_trait::async_trait;
use candid::CandidType;
use derive_more::{Display, From};

#[derive(Debug, Display, From)]
pub enum Error {}

pub trait Event: CandidType + Send + Sync + Clone + 'static {
    /// Create an event from a log.
    fn from<T>(log: T) -> Self;
}

#[async_trait]
pub trait Indexer {
    /// Get events from `from` to `to`. e.g. `from = 1`, `to = 10` will return events with id from 1 to 10.
    async fn get_from_to<U>(&self, from: u64, to: u64) -> Result<HashMap<u64, Vec<U>>, Error>;
    /// Save events to database.
    fn save<T>(&self, id: u64, elements: Vec<T>)
    where
        T: Event;
    /// Get the latest event id.
    fn get_last_number(&self) -> Result<u64, Error>;
    /// Get the event chunk size.
    fn get_event_chunk_size(&self) -> Result<u64, Error>;
    /// Set the event chunk size.
    fn set_event_chunk_size(&self, size: u64) -> Result<(), Error>;
    /// Get the last indexed event id.
    fn get_last_indexed(&self) -> Result<u64, Error>;
    /// Set the last indexed event id.
    fn set_last_indexed(&self, id: u64) -> Result<(), Error>;
    /// Index events.
    async fn index<T, U>(&self) -> Result<(), Error>
    where
        T: Event,
    {
        let last_indexed = self.get_last_indexed()?;
        let chunk_size = self.get_event_chunk_size()?;
        let from = last_indexed + 1;
        let to = last_indexed + chunk_size;
        self.get_from_to(from, to)
            .await?
            .into_iter()
            .map(|(k, v)| (k, v.into_iter().map(Event::from::<U>).collect::<Vec<T>>()))
            .for_each(|(k, v)| {
                self.save(k, v);
            });
        self.set_last_indexed(to)?;
        Ok(())
    }
}
