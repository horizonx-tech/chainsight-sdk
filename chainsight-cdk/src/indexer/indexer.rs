use std::collections::HashMap;

use async_trait::async_trait;
use candid::CandidType;
use derive_more::Display;
use serde::Deserialize;

use crate::storage::{self, Data};

#[derive(Debug, Display)]
pub enum Error {
    #[display(fmt = "Indexer error: {}", _0)]
    IndexerError(String),
    #[display(fmt = "Web3 error: {}", _0)]
    Web3Error(String),
    #[display(fmt = "Database error: {}", _0)]
    DatabaseError(String),
    #[display(fmt = "Other error: {}", _0)]
    OtherError(String),
}

#[derive(CandidType, Clone, Debug, Default, PartialEq, Deserialize)]
pub struct IndexingConfig {
    pub start_from: u64,
    pub chunk_size: Option<u64>,
}

pub trait Event<T>: CandidType + Send + Sync + Clone + 'static {
    fn from(event: T) -> Self
    where
        T: Into<Self>;
    fn tokenize(&self) -> Data;
    fn untokenize(data: Data) -> Self;
}

#[async_trait]
pub trait Finder<T>: Send + Sync + Clone + 'static {
    /// Get events from `from` to `to`. e.g. `from = 1`, `to = 10` will return events with id from 1 to 10.
    async fn find(&self, from: u64, to: u64) -> Result<HashMap<u64, Vec<T>>, Error>;
}

#[async_trait]
pub trait Indexer<T> {
    type Finder: Finder<T>;
    fn finder(&self) -> Self::Finder;
    /// Save events to database.
    fn save<E>(&self, id: u64, elements: Vec<E>)
    where
        E: Event<T>,
    {
        elements.iter().map(|e| e.tokenize()).for_each(|data| {
            storage::insert(id, data);
        });
    }
    fn between<E>(&self, from: u64, to: u64) -> Result<HashMap<u64, Vec<E>>, Error>
    where
        E: Event<T>,
    {
        let mut res = HashMap::new();
        storage::between(from, to).into_iter().for_each(|e| {
            let data =
                e.1.to_vec()
                    .iter()
                    .map(|e| Event::untokenize(e.clone()))
                    .collect::<Vec<E>>();
            res.insert(e.0, data);
        });
        Ok(res)
    }
    /// Get the last indexed event id.
    fn get_last_indexed(&self) -> Result<u64, Error> {
        Ok(storage::last(1).iter().map(|e| e.0).max().unwrap_or(0))
    }
    /// Index events.
    async fn index<E>(&self, cfg: IndexingConfig) -> Result<(), Error>
    where
        E: Event<T> + From<T>,
    {
        let last_indexed = self.get_last_indexed()?;
        let chunk_size = cfg.chunk_size.unwrap_or(500);
        let from = cfg.start_from.max(last_indexed + 1);
        let to = from + chunk_size;
        let elements: HashMap<u64, Vec<T>> = self.finder().find(from, to).await?;
        self.on_update::<E>(elements);
        Ok(())
    }

    fn on_update<E>(&self, elements: HashMap<u64, Vec<T>>)
    where
        E: Event<T> + From<T>,
    {
        elements
            .into_iter()
            .map(|(k, v)| (k, v.into_iter().map(Event::from).collect::<Vec<E>>()))
            .for_each(|(k, v)| {
                self.save(k, v);
            });
    }
}
