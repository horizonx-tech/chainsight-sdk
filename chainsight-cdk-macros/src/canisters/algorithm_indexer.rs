use chainsight_cdk::config::components::{AlgorithmIndexerConfig, AlgorithmInputType};
use proc_macro::TokenStream;
use quote::{format_ident, quote};
use syn::parse_macro_input;

pub fn def_algorithm_indexer_canister(input: TokenStream) -> TokenStream {
    let input_json_string = parse_macro_input!(input as syn::LitStr).value();
    let config: AlgorithmIndexerConfig =
        serde_json::from_str(&input_json_string).expect("Failed to parse input_json_string");
    algorithm_indexer_canister(config).into()
}

fn algorithm_indexer_canister(config: AlgorithmIndexerConfig) -> proc_macro2::TokenStream {
    let AlgorithmIndexerConfig {
        common,
        indexing: _,
        input,
    } = config;
    let monitor_duration = common.monitor_duration;
    let canister_name = common.canister_name.clone();
    let canister_name_ident = format_ident!("{}", common.canister_name);
    let input_ty = input_type_ident(input.response_type, input.source_type);

    let method_name = input.method_name;
    quote! {
        use candid::CandidType;
        use chainsight_cdk::indexer::IndexingConfig;
        use chainsight_cdk_macros::{
            algorithm_indexer, chainsight_common, did_export, init_in, manage_single_state, setup_func,
            timer_task_func,
        };
        use serde::{Deserialize, Serialize};
        use std::collections::HashMap;
        chainsight_common!(#monitor_duration);
        init_in!();
        manage_single_state!("target_addr", String, false);
        setup_func!({ target_addr: String, config: IndexingConfig });
        timer_task_func!("set_task", "index");
        use #canister_name_ident::*;

        algorithm_indexer!(#input_ty, #method_name);
        did_export!(#canister_name);
    }
}

pub fn input_type_ident(
    struct_name: String,
    source_type: AlgorithmInputType,
) -> proc_macro2::TokenStream {
    let event_struct = format_ident!("{}", &struct_name);
    match source_type {
        AlgorithmInputType::EventIndexer => {
            // HashMap<u64, Vec<event_struct>>
            let source_ident = format_ident!("{}", &"HashMap".to_string());
            quote! {
                #source_ident<u64, Vec<#event_struct>>
            }
        }
        AlgorithmInputType::KeyValue => {
            // HashMap<String, event_struct>
            let source_ident = format_ident!("{}", &"HashMap".to_string());
            quote! {
                #source_ident<String, #event_struct>
            }
        }
        AlgorithmInputType::KeyValues => {
            // HashMap<String, Vec<event_struct>>
            let source_ident = format_ident!("{}", &"HashMap".to_string());
            quote! {
                #source_ident<String, Vec<#event_struct>>
            }
        }
    }
}

#[cfg(test)]
mod test {
    use chainsight_cdk::{
        config::components::{AlgorithmIndexerInput, CommonConfig},
        indexer::IndexingConfig,
    };
    use insta::assert_snapshot;
    use rust_format::{Formatter, RustFmt};

    use super::*;

    #[test]
    fn test_snapshot() {
        let config = AlgorithmIndexerConfig {
            common: CommonConfig {
                monitor_duration: 3600,
                canister_name: "example_canister".to_string(),
            },
            indexing: IndexingConfig {
                start_from: 1222222,
                chunk_size: None,
            },
            input: AlgorithmIndexerInput {
                method_name: "get_list".to_string(),
                response_type: "String".to_string(),
                source_type: AlgorithmInputType::EventIndexer,
            },
        };
        let generated = algorithm_indexer_canister(config);
        let formatted = RustFmt::default()
            .format_str(generated.to_string())
            .expect("rustfmt failed");
        assert_snapshot!("snapshot__algorithm_indexer", formatted);
    }
}
