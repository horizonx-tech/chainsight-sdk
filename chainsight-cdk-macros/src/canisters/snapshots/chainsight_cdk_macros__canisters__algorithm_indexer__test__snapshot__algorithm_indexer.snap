---
source: chainsight-cdk-macros/src/canisters/algorithm_indexer.rs
expression: formatted
---
use candid::{CandidType, Decode, Encode};
use chainsight_cdk::indexer::IndexingConfig;
use chainsight_cdk_macros::{
    algorithm_indexer, algorithm_indexer_source, chainsight_common, did_export,
    generate_queries_for_key_value_store_struct, generate_queries_for_key_values_store_struct,
    init_in, manage_single_state, prepare_stable_structure, setup_func, stable_memory_for_scalar,
    timer_task_func, CborSerde, StableMemoryStorable,
};
use ic_stable_structures::writer::Writer;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
did_export!("example_canister");
chainsight_common!();
init_in!(11);
prepare_stable_structure!();
stable_memory_for_scalar!("target_addr", String, 12, false);
setup_func ! ({ target_addr : String , config : IndexingConfig } , 13);
timer_task_func!("set_task", "index", 14);
use example_canister::*;
algorithm_indexer_source!();
algorithm_indexer ! (HashMap < u64 , Vec < String >> , "get_list" , 15);
generate_queries_for_key_value_store_struct!(OutputType1);
generate_queries_for_key_values_store_struct!(OutputType2);

